Add iOS SDK 4.19.0 changelog and reference link [SPEX-2323] (#231)
MapsPeople/documentation: other/changelog/ios-sdk/v4.md

{% include "../../../.gitbook/includes/ios-xcode-16-requirement.md" %}

### \[4.19.0] 2026-07-30

#### Added

> * Route styling is now configured through a single `MPDirectionsRendererOptions` object on `MPDirectionsRenderer.options`. It covers the route line — colour, opacity, weight, and a solid, dashed or dotted stroke style — an optional halo around it, and a separately styled overlay that animates along the route as `flow`, `pulse` or `comet` (or `none`). It also adds repeating direction indicators along the line, from five built-in arrow shapes or a custom image; custom start and destination markers with their own icons and labels; per-connector-type icons at floor changes; a cap on how far a route fit zooms in; and a 3D elevated line \[Mapbox only]. Route animation respects the system Reduce Motion setting unless `forceAnimation` is set. The same configuration can also be set per solution in the CMS and is applied automatically, exposed as `MPSolutionConfig.directionsRendererConfig`, with an option set in code taking precedence over the solution value, and the solution value over the SDK default.

Reviewer: This single sentence-chain carries about ten distinct features. As written, our biggest release in months is unscannable and unlinkable. I suggest a lead-in bullet naming `MPDirectionsRendererOptions` plus sub-bullets per capability, with the CMS precedence rule as its own bullet.

```suggestion
* Route styling is now configured through a single `MPDirectionsRendererOptions` object on `MPDirectionsRenderer.options`:
  * Route line styling: color, opacity, weight, and a solid, dashed or dotted stroke style.
  * An optional halo drawn around the route line.
  * A separately styled overlay that animates along the route as `flow`, `pulse` or `comet` (or `none`).
  * Repeating direction indicators along the line, from five built-in arrow shapes or a custom image.
  * Custom start and destination markers with their own icons and labels.
  * Per-connector-type icons at floor changes.
  * A cap on how far a route fit zooms in.
  * A 3D elevated route line \[Mapbox only].
  * Route animation respects the system Reduce Motion setting unless `forceAnimation` is set.
  * The same configuration can be set per solution in the CMS and is applied automatically, exposed as `MPSolutionConfig.directionsRendererConfig`.
  * Precedence when the same setting is defined in more than one place: options set in code override `MPSolutionConfig.directionsRendererConfig`, which overrides the SDK default.
```

Author: Applied as suggested.

* Added step-level route navigation alongside the existing leg-level API: `routeStepIndex`, `nextStep()` and `previousStep()`, with an `onStepSelected(stepIndex:)` delegate callback. A route can also be shown and hidden with `isRouteVisible`, and the leg buttons toggled while a route is displayed with `showRouteLegButtons`.

> * Added `MPMapControl.userPositionOpacityOverride` to opt out of dimming the user's position when it is on a floor other than the one displayed. Set a value to apply that opacity on every floor; leave it unset to keep the current behaviour.

Reviewer: We have standardized on US English, so

```suggestion
* Added `MPMapControl.userPositionOpacityOverride` to opt out of dimming the user's position when it is on a floor other than the one displayed. Set a value to apply that opacity on every floor; leave it unset to keep the current behavior.
```

Author: Done.

#### Changed

> * The default route appearance has changed. Previously the route drew a wide line at 30% opacity with a narrower full-colour line animating along it; it now draws a solid line with a thinner, lighter overlay travelling along it, and the halo around the line is off by default. Set the corresponding `MPDirectionsRendererOptions` properties to restore a previous look.

Reviewer: We have standardized on US English, so

```suggestion
* The default route appearance has changed. Previously the route drew a wide line at 30% opacity with a narrower full-color line animating along it; it now draws a solid line with a thinner, lighter overlay traveling along it, and the halo around the line is off by default. Set the corresponding `MPDirectionsRendererOptions` properties to restore a previous look.
```

Author: Done.

* Redrawing a route that is already shown — for example on a floor change, or when changing leg — no longer re-sends the route rendering start and stop notifications or the corresponding analytics event. Setting `routeLegIndex` to the value it already holds no longer triggers `onLegSelected`.

> * Data providers now serve from an in-memory cache before going to the network, removing redundant requests within a session. The cache is cleared when a solution is loaded or synchronised.

Reviewer: We have standardized on US English, so

```suggestion
* Data providers now serve from an in-memory cache before going to the network, removing redundant requests within a session. The cache is cleared when a solution is loaded or synchronized.
```

Author: Done.

> * Updated the Google Maps SDK to 10.15.0 (from 9.4.0) and the Mapbox Maps SDK to 11.25.0 (from 11.18.2). The Swift Package for Google Maps now resolves an exact Google Maps version rather than a range.

Reviewer: No need to mention there was a range before – that was a one off, and the customer that requested it had to support iOS 15 devices which we don't any longer…

```suggestion
* Updated the Google Maps SDK to 10.15.0 (from 9.4.0) and the Mapbox Maps SDK to 11.25.0 (from 11.18.2).
```

Author: Removed.

#### Deprecated

> * `MPDirectionsRenderer.fitBounds`, `padding`, `pathColor` and `backgroundColor` are deprecated in favour of `fitBounds`, `fitBoundsPadding`, `strokeColor` and `backgroundColor` on `MPDirectionsRendererOptions`. The existing properties continue to work.

Reviewer: We have standardized on US English, so

```suggestion
* `MPDirectionsRenderer.fitBounds`, `padding`, `pathColor` and `backgroundColor` are deprecated in favor of `fitBounds`, `fitBoundsPadding`, `strokeColor` and `backgroundColor` on `MPDirectionsRendererOptions`. The existing properties continue to work.
```

Author: Done.

#### Fixed

* Fixed a crash caused by concurrent access while rendering on Google Maps.
* Fixed data races in the shared native library.
* Fixed data races in the route flow animation, which now runs on a native display-synchronised animator.
* Filtered searches no longer risk blocking a thread while converting filters, which could stall other concurrent work.

> * Route camera fits now honour the solution's automated zoom limit on Google Maps, which previously ignored it. Mapbox was unaffected.

Reviewer: We have standardized on US English, so

```suggestion
* Route camera fits now honor the solution's automated zoom limit on Google Maps, which previously ignored it. Mapbox was unaffected.
```

Author: Done.
